src/Voron/Data/Tables/TableValueCompressor.cs

```diff
@@ -0,0 +1,401 @@
+            {
+                if (it.SeekToLast() == false)
+                    return false;
+
+                long lastEtag = it.CurrentKey;
+                long total  = etagTree.NumberOfEntries;
+
+                if(total < 16*1024)
+                    return (lastEtag & 1024) == 0;
+                if(total < 64 * 1024)
+                    return (lastEtag & 2048) == 0;
+                if(total < 256*1024)
+                    return (lastEtag & 8192) == 0;
+                if (total < 512 * 1024)
+                    return (lastEtag & 16 * 1024) == 0;
+                return (lastEtag & 32 * 1024) == 0;
+            }
+        }
+
+        private byte GetCompressionRatio(int compressed, int raw)
+        {
+            return (byte)(compressed / (float)raw * 100);
+        }
+
+        public int WriteVariableSizeIntInReverse(byte* buffer, int value)
+        {
+            var count = 0;
+            var v = (uint)value;
+            while (v >= 0x80)
+            {
+                buffer[count++] = (byte)(v | 0x80);
+                v >>= 7;
+            }
+
+            buffer[count++] = (byte)v;
+
+            for (int i = count - 1; i >= count / 2; i--)
+            {
+                var tmp = buffer[i];
+                buffer[i] = buffer[count - 1 - i];
+                buffer[count - 1 - i] = tmp;
+            }
+
+            return count;
+        }
+
+
+        private void MaybeTrainCompressionDictionary(Table table, FixedSizeTree etagsTree)
+        {
+            // the idea is that we'll get better results by including the most recently modified documents
+            // by iterating over the tag index, which is guaranteed to be always increasing
+            var dataIds = new List<long>(256);
+            var sizes = new UIntPtr[256];
+            var totalSize = 0;
+            int totalSkipped =0 ;
+            using (var it = etagsTree.Iterate())
+            {
+                if (it.SeekToLast() == false)
+                    return; // empty table, nothing to train on
+
+                do
+                {
+                    long id = it.CreateReaderForCurrent().ReadLittleEndianInt64();
+                    table.DirectRead(id, out var size);
+                    if (size > 32 * 1024)
+                    {
+                        if (totalSkipped++ > 16 * 1024)
+                            return;  // we are scanning too much, no need to try this hard
+                        // we don't want to skip documents that are too big, they will compress
+                        // well on their own, and likely be *too* unique to add meaningfully to the
+                        // dictionary
+                        continue; 
+                    }
+
+                    sizes[dataIds.Count] = (UIntPtr)size;
+                    dataIds.Add(id);
+                    totalSize += size;
+                } while (dataIds.Count < 256 && it.MovePrev() && totalSize < 1024 * 1024);
```

Comment:
What is the reasoning that we iterate from the last element?

Reply:
We want to get the most recently update values, we assume that those are the most similar to the currently inserted dta.